### ISSUE TITLE
Packaging broken on Alpine Linux: Error: /bin/sh: ./pact-provider-verifier: not found

### ISSUE THREAD

**Issue by emk** [pact-foundation/pact-provider-verifier#16]
At work, most of our services run inside Docker containers, and most of those Docker containers are based on [Alpine Linux](https://alpinelinux.org/). Alpine Linux is a stripped down Linux distro, with a typical base size between (IIRC) 10 and 35 MB depending on what version and tools you're using. This speeds up container deployments.

Alpine Linux doesn't actually contain a copy of `glibc`. Instead, it uses a copy of [musl-libc](https://www.musl-libc.org/), a lightweight embedded C library.

But this means that applications which are dynamically linked against `glibc` will fail to run with a `not found` error:

```
  not ok 1 Error: /bin/sh: ./pact-provider-verifier: not found 
    ---
      operator: error
      expected: |-
        undefined
      actual: |-
        [Error: /bin/sh: ./pact-provider-verifier: not found
  ]
      at: process._tickDomainCallback (internal/process/next_tick.js:228:7)
      stack: |-
        Error: /bin/sh: ./pact-provider-verifier: not found
```

## Possible fixes

There are a number of different plausible approaches to making Pact verification work on Alpine Linux:

1. Make it easier to use the system Ruby interpreter, perhaps using an environment variable to configure it.
2. Statically link the bundled `ruby` on Linux using `musl-libc`. This is tricky, but somebody may have already done it. This will run on _any_ Linux distro.
3. Use a single, statically-linked Rust binary. I think some work towards this has already been done, and I would be happy to help you set up 100% static linking for Rust binaries. (OpenSSL makes this a little tricky, but I know [the workarounds](https://github.com/emk/rust-musl-builder) and I'm happy to help set them up.)

Any of these would allow us to use Pact on many more of our internal projects.

## Testing whether a binary relies on shared libraries

To check whether a binary relies on shared libraries, you can use `ldd`. Here's a dynamically-linked Ruby:

```
$ ldd /usr/bin/ruby
	linux-vdso.so.1 =>  (0x00007fff0a7cb000)
	libruby-2.3.so.2.3 => /usr/lib/x86_64-linux-gnu/libruby-2.3.so.2.3 (0x00007fe165795000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe1653ce000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fe1651b0000)
	libgmp.so.10 => /usr/lib/x86_64-linux-gnu/libgmp.so.10 (0x00007fe164f30000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fe164d2c000)
	libcrypt.so.1 => /lib/x86_64-linux-gnu/libcrypt.so.1 (0x00007fe164af2000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007fe1647e9000)
	/lib64/ld-linux-x86-64.so.2 (0x000055b9e9eb6000)
```

Here's a statically-linked one:

```
$ ldd /usr/bin/ruby
	not a dynamic executable
```

**Comment by mefellows** [comment]
I think this has been solved somewhere already, I'll see if I can find the implementation.

Once rust is ready and at feature parity we should move over to it.

**Comment by bethesque** [comment]
How about you use this? https://github.com/DiUS/pact-provider-verifier-docker

**Comment by mefellows** [comment]
Yes, that was what I was thinking of :)

**Comment by bethesque** [comment]
Closing.
